Helix Restore Issue Mitigation

TanayParikh opened the pull request.

TanayParikh committed:
RunTests dotnet run --no-build

TanayParikh committed:
Remove duplicate run

TanayParikh committed:
Add to Build.props

dougbu commented on eng/Build.props:
```diff
@@ -240,6 +241,7 @@
                           $(RepoRoot)src\Extensions\**\src\*.csproj;
                           $(RepoRoot)src\BuildAfterTargetingPack\*.csproj;
                           $(RepoRoot)src\OpenApi\**\src\*.csproj;
+                          $(RepoRoot)eng\helix\content\RunTests\RunTests.csproj;
```
Don't add the project here

dougbu commented on eng/helix/content/runtests.cmd:
```diff
@@ -23,15 +23,6 @@ echo.
-)
-
 echo "Running tests: dotnet run --no-restore --project RunTests\RunTests.csproj -- --target %$target% --runtime %$aspRuntimeVersion% --queue %$queue% --arch %$arch% --quarantined %$quarantined% --ef %$ef% --helixTimeout %$helixTimeout%"
 dotnet run --no-restore --project RunTests\RunTests.csproj -- --target %$target% --runtime %$aspRuntimeVersion% --queue %$queue% --arch %$arch% --quarantined %$quarantined% --ef %$ef% --helixTimeout %$helixTimeout%
```
Add `--no-build` here instead of `--no-restore`

dougbu commented on eng/helix/content/runtests.sh:
```diff
@@ -75,18 +75,6 @@ sync
-fi
-
 echo "Running tests: dotnet run --no-restore --project RunTests/RunTests.csproj -- --target $1 --runtime $2 --queue $helixQueue --arch $4 --quarantined $5 --ef $6 --helixTimeout $7"
 dotnet run --no-restore --project RunTests/RunTests.csproj -- --target $1 --runtime $2 --queue $helixQueue --arch $4 --quarantined $5 --ef $6 --helixTimeout $7
```
Same here

TanayParikh committed:
Treat CA2007 as a warning in RunTests

TanayParikh committed:
NoWarn CA2007

dougbu commented on eng/helix/content/RunTests/RunTests.csproj:
```diff
@@ -4,6 +4,7 @@
     <OutputType>Exe</OutputType>
     <TargetFramework>net7.0</TargetFramework>
     <DefineConstants Condition=" '$(InstallPlaywright)' == 'true' ">$(DefineConstants);INSTALLPLAYWRIGHT</DefineConstants>
+    <NoWarn>$(NoWarn);CA2007</NoWarn>
```
why not add a `ConfigureAwait(...)`❔

TanayParikh replied:
Figured it didn't really matter in this context (tests). 

We do it elsewhere, as well:

https://github.com/dotnet/aspnetcore/blob/f041d49f3320fd0fb70babe35a861a9035e48e7e/eng/targets/CSharp.Common.targets#L37-L41

dougbu replied:
Shouldn't need to duplicate that 😸

TanayParikh replied:
> Would still prefer you remove the duplication around `CA2007`

https://github.com/dotnet/aspnetcore/pull/41311/commits/3858e46da674aee597ae39c1925680b6ac5c3eb8

TanayParikh committed:
Ignore NU5104 for System.CommandLine prerelease package

TanayParikh committed:
PR Feedback

dougbu reviewed (approved):
If CI validation works, great. Would still prefer you remove the duplication around `CA2007`

TanayParikh commented:
```
/mnt/vss/_work/1/s/eng/helix/content/RunTests/RunTests.csproj : error MSB4057: The target "Test" does not exist in the project.
##[error]eng/helix/content/RunTests/RunTests.csproj(0,0): error MSB4057: (NETCORE_ENGINEERING_TELEMETRY=Build) The target "Test" does not exist in the project.
```

https://dev.azure.com/dnceng/public/_build/results?buildId=1731724&view=logs&jobId=5ac7b393-e840-5549-7fb4-a4479af8e7e3&j=5ac7b393-e840-5549-7fb4-a4479af8e7e3&t=29df2fa2-0d20-51bd-e85a-8b546e86c529

dougbu commented:
@TanayParikh to avoid that, set `$(IsUnitTestProject)` and `$(IsTestProject)` to `false` in the project. Wasn't an issue before but this project name matches our usual test project names

dougbu commented on eng/helix/content/RunTests/ProcessUtil.cs:
```diff
@@ -171,17 +170,17 @@ public static Task CaptureDumpAsync(int pid, string dumpFilePath)
 
             var canceledTcs = new TaskCompletionSource<object?>();
-            await using var _ = cancellationToken.Register(() => canceledTcs.TrySetResult(null));
+            await using var _ = cancellationToken.Register(() => canceledTcs.TrySetResult(null)).ConfigureAwait(false);
```
I suggested adding `ConfigureAwait(...)` only because I thought the number of cases was low. Seeing this, I prefer adding `CA2007`. That's necessary because the project targets `net7.0` and not anything `netstandard`

dougbu reviewed (approved):
🥇

dougbu commented:
I still approve 😃

TanayParikh commented:
> @TanayParikh to avoid that, set `$(IsUnitTestProject)` and `$(IsTestProject)` to `false` in the project. Wasn't an issue before but this project name matches our usual test project names

Hmm that didn't seem to do it, the issue is persisting: https://dev.azure.com/dnceng/public/_build/results?buildId=1731849&view=logs&jobId=5ac7b393-e840-5549-7fb4-a4479af8e7e3&j=5ac7b393-e840-5549-7fb4-a4479af8e7e3&t=29df2fa2-0d20-51bd-e85a-8b546e86c529

https://github.com/dotnet/aspnetcore/pull/41311/commits/5a96d33b3514949785a7c754b596523f8cc8c4e4 set `IsUnitTestProject` and `IsTestProject` to `false`

dougbu commented on eng/helix/content/RunTests/RunTests.csproj:
```diff
@@ -4,6 +4,9 @@
     <TargetFramework>net7.0</TargetFramework>
     <DefineConstants Condition=" '$(InstallPlaywright)' == 'true' ">$(DefineConstants);INSTALLPLAYWRIGHT</DefineConstants>
+    <NoWarn>$(NoWarn);CA2007;NU5104</NoWarn>
+    <IsUnitTestProject>false</IsUnitTestProject>
+    <IsTestProject>false</IsTestProject>
```
Arcade doesn't work the way I'd hoped and overrides `$(IsTestProject)`. Instead of this and `Sdk="Microsoft.NET.Sdk"` in the `<Project>` tag:

``` msbuild
  <PropertyGroup>
    <IsUnitTestProject Condition=" '$(SkipTestBuild)' != 'true' ">true</IsUnitTestProject>
  </PropertyGroup>

  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
```

Will also need `<Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />` near the bottom of the file.

TanayParikh replied:
94105ca should handle that.

dougbu commented on eng/helix/content/RunTests/RunTests.csproj:
```diff
@@ -1,13 +1,21 @@
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <IsUnitTestProject Condition=" '$(SkipTestBuild)' != 'true' ">true</IsUnitTestProject>
```
This is the opposite of what you want. `$(IsUnitTestProject)` defaults to `true` and that's causing Arcade to attempt to run tests in this project.

nit: Only this `<PropertyGroup/>` needs to be above the Sdk.props `<Import/>`

TanayParikh replied:
> opposite of what you want

```
<IsUnitTestProject Condition=" '$(SkipTestBuild)' != 'true' ">false</IsUnitTestProject>
```

is what's intended?

dougbu replied:
I wouldn't bother w/ the `Condition` but that's not a big deal at all. Otherwise fine.

TanayParikh replied:
[576767a](https://github.com/dotnet/aspnetcore/pull/41311/commits/576767a7c5ca025db8d629ec47c7d5874eba566c)

TanayParikh replied:
Still the same thing?

> eng/helix/content/RunTests/RunTests.csproj(0,0): error MSB4057: (NETCORE_ENGINEERING_TELEMETRY=Build) The target "Test" does not exist in the project.

dougbu replied:
Opened a binary log from this last build and see `Target "Test" skipped, due to false condition; ('$(IsUnitTestProject)' == 'true' or '$(IsPerformanceTestProject)' == 'true') was evaluated as ('false' == 'true' or 'false' == 'true').` for adjacent projects. That indicates the `Test` target is successfully defined by Arcade's Tests.targets file elsewhere. 

OIC 🤦: eng/helix/ contains Directory.Build.* files that prevent `dotnet msbuild` seeing our root Directory.Build.* files, meaning the Arcade SDK doesn't get imported. Simple fix is to revert everything related to `$(IsTestProject)`, `$(IsUnitTestProject)`, go back to the regular .NET SDK approach, and add `<Target Name="Test" />` to the project.

Sorry again for my unhelpfulness here. Few too many things going on I guess&hellip;

TanayParikh committed:
PR Feedback

dougbu committed:
Generate .config/dotnet-tools.json
  - add generation to GenerateFiles.csproj
  - include all required versions in eng/Versions.props
    - nit: bump tool versions slightly
    - `dotnet-dump` move from `5.0.0-*` to `6.0.322601` is largest version bump
  - have `git` ignore generated file
    - nit: put `*.svclog` together w/ other extension exclusions

dougbu committed:
Do not restore tools from network
  - get tool packages from Helix correlation root
  - to do this, save and restore NuGet.config file
  - this removes `--version` from `dotnet-dump` and `dotnet-ef` installations
    - will only have a single package for each tool in the correlation payload

dougbu committed:
Do not pass `dotnet-ef` version to `RunTests`
  - mostly cleanup; no longer needed

dougbu committed:
Add tool packages to Helix correlation payloads

dougbu committed:
Remove DotNetToolsInstaller.csproj
  - mostly cleanup
  - `dotnet-ef` tool now restored by Arcade's Tools.proj much earlier in our build because
    most configured tool packages are needed in `RunTests` on Helix agents

dougbu committed:
Add `--playwright` option in `Runtests`
  - remove `INSTALLPLAYWRIGHT` define and `$env:INSTALLPLAYWRIGHT`
  - always reference Microsoft.Playwright in `RunTests`
  - nit: `InstallPlaywrightAsync()` wasn't `async`; fix it and rename to `InstallPlaywright()`

dougbu committed:
Use Arcade SDK in `RunTests` and `Publish` project
  - match most other projects in this repo
    - remove empty Directory.Build.props and .targets files preventing Arcade imports
  - exclude project build if `$(SkipTestBuild)` (though not a test project)

dougbu committed:
Correct minor problems in `RunTests` source

dougbu committed:
Use file-scoped namespaces in `RunTests` source

dougbu committed:
Use VS code fixes in `RunTests`

dougbu committed:
Use published `RunTests` output on Helix agents

dougbu committed:
Move `RunTests` under eng/tools/
- don't need the project on Helix agents
- restore Directory.Build.* files removed when switching to Arcade

dougbu committed:
Add `RunTests` to tools.slnf

dougbu commented:
I rewrote the history of this PR to remove reverted commits, rebased on 'main', and added the workarounds discussed [above](https://github.com/dotnet/aspnetcore/pull/41311#discussion_r856595237) and in Teams. The result seems simpler than I expected though the 11 brand new commits were a bit of work 😉

Some of the commits are strictly cleanup as I went along (hitting issues).

dougbu commented on eng/tools/RunTests/ProcessResult.cs:
```diff
@@ -0,0 +1,18 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Due to folder move and file-scoped namespace change, it's much easier to look at individual commits to see what changed here.

wtgodbe commented on eng/helix/helix.proj:
```diff
@@ -47,6 +47,15 @@
+    <HelixCorrelationPayload Include="$(NUGET_PACKAGES)\dotnet-serve\$(DotnetServeVersion)\dotnet-serve.$(DotnetServeVersion).nupkg" />
+
+    <!-- Grab published `RunTests` project output. -->
+    <HelixCorrelationPayload Include="$(ArtifactsBinDir)RunTests\Debug\net7.0\publish\"
```
Can you parameterize `net7.0`?

dougbu replied:
Due to the special cases we hit when moving to new TFMs, no. I did however add a comment about keeping things in sync in RunTests.csproj

HaoK commented on eng/helix/content/RunTests/Program.cs:
```diff
@@ -19,14 +19,18 @@ static async Task Main(string[] args)
                     keepGoing = await runner.InstallDotnetToolsAsync();
                 }
-#if INSTALLPLAYWRIGHT
+
```
I believe the reason I did this was because some queues don't support playwright but maybe now that we are building this on CI rather than on the actual machines, this doesn't matter anymore, but just something to be aware of

HaoK commented:
Looks great overall, I'd suggest we take this opportunity now that we are moving it to rename this RunHelixTests or something similar since its not longer in the helix content directory

HaoK commented:
HelixTestRunner for the project is my vote

dougbu committed:
Rename `RunTests` project to `HelixTestRunner`
- less confusing given `RunTests` target and runtests.sh et cetera

dougbu commented:
> HelixTestRunner for the project is my vote

Good suggestion, especially since we hook into an Arcade `RunTests` target and have scripts named `runtests.*`

dougbu committed:
!fixup! Define `$(ArtifactsBinDir)` in helix.proj

dougbu committed:
!fixup! Use `$(Configuration)` in helix.proj

HaoK reviewed (approved):
Nice!

HaoK commented on eng/tools/GenerateFiles/GenerateFiles.csproj:
```diff
@@ -15,10 +17,14 @@
         AspNetCorePatchVersion=$(AspNetCorePatchVersion);
         DefaultNetCoreTargetFramework=$(DefaultNetCoreTargetFramework);
+        DotnetDumpVersion=$(DotnetDumpVersion);
+        dotnetefVersion=$(dotnetefVersion);
```
minor nit: casing of dotnetefVersion here is wonky compared to everything else like DotnetDump

dougbu replied:
I agree but it has to match the casing of the dependency in Version.Details.xml for strange reasons related to the wild logic in the https://github.com/dotnet/aspnetcore/blob/main/eng/Dependencies.props#L218 item group.

I may see if property name case-insensitivity makes a difference these days but will leave that for another PR (even though this one is failing validation).

HaoK commented on eng/helix/helix.proj:
```diff
@@ -47,6 +49,15 @@
+    <HelixCorrelationPayload Include="$(NUGET_PACKAGES)\dotnet-serve\$(DotnetServeVersion)\dotnet-serve.$(DotnetServeVersion).nupkg" />
+
+    <!-- Grab published `HelixTestRunner` project output. -->
+    <HelixCorrelationPayload Include="$(ArtifactsBinDir)HelixTestRunner\$(Configuration)\net7.0\publish\"
```
Maybe easier to just copy the publish directory into the workitem rather than using correlationpayloads for the helix test runner, its not like a huge publish dir right?  Especially if the same dll would work on all queues since I'm assuming we don't have any plat specific code in the runner

dougbu replied:
I disagree because including the files in `<HelixContent />` embiggens hundreds of work item payloads rather than the 3-11 correlation payloads. This also keeps things together w/ the other platform-neutral / non-test-specific bits like the tool packages above and the App.Ref / App.Runtime (non-`crossgen`ed) payloads near the end of this file. I also don't see anything "simpler" about your proposal. Am I missing something❔

----

In fact, I'm tempted to move all but `@(HelixContent)` from https://github.com/dotnet/aspnetcore/blob/f6cddc1136fa9310a8fa0a06af8a461b409071a9/eng/targets/Helix.targets#L79-L91 into a correlation payload if only to reduce the number of times those small files are duplicated in `publish` directories and then work item payloads. Not an immediate problem but does cost a bit.

Somewhat separately there's no reason not to move https://github.com/dotnet/aspnetcore/blob/f6cddc1136fa9310a8fa0a06af8a461b409071a9/eng/targets/Helix.props#L27 into the correlation payloads.

HaoK replied:
Okay I leave it to you, this part of the infrastructure has evolved very far from where it started

dougbu committed:
!fixup! Fix typo in property function

dougbu committed:
!fixup! Fix `dotnet` syntax in runtests.sh

dougbu commented:
Found what I think is the last bug in this PR. `%(HelixCorrelationPayload.Destination)` metadata shouldn't have a trailing slash of any type. Adding a backslash causes problems on Unix-based Helix agents.

![image](https://user-images.githubusercontent.com/6431421/166082301-931bc1bd-0b3d-444a-9b77-1c1a9d595a50.png)

dougbu committed:
!fixup! Correct `%(HelixCorrelationPayload.Destination)`
